tests/ga/test_collect_logs.py

```diff
@@ -0,0 +1,187 @@
+                        with patch("azurelinuxagent.ga.collect_logs.conf.get_collect_logs", return_value=True):
+                            def run_and_wait():
+                                collect_logs_handler.run()
+                                collect_logs_handler.join()
+
+                            collect_logs_handler = get_collect_logs_handler()
+                            collect_logs_handler.get_mock_wire_protocol = lambda: protocol
+                            collect_logs_handler.run_and_wait = run_and_wait
+                            yield collect_logs_handler
+
+
+@skip_if_predicate_true(is_python_version_26, "Disabled on Python 2.6")
+class TestCollectLogs(AgentTestCase, HttpRequestPredicates):
+    def setUp(self):
+        AgentTestCase.setUp(self)
+        prefix = "UnitTest"
+        logger.DEFAULT_LOGGER = Logger(prefix=prefix)
+
+        self.archive_path = os.path.join(self.tmp_dir, "logs.zip")
+        self.mock_archive_path = patch("azurelinuxagent.ga.collect_logs.COMPRESSED_ARCHIVE_PATH", self.archive_path)
+        self.mock_archive_path.start()
+
+        # Since ProtocolUtil is a singleton per thread, we need to clear it to ensure that the test cases do not
+        # reuse a previous state
+        clear_singleton_instances(ProtocolUtil)
+
+    def tearDown(self):
+        if os.path.exists(self.archive_path):
+            os.remove(self.archive_path)
+        self.mock_archive_path.stop()
+        AgentTestCase.tearDown(self)
+
+    def _create_dummy_archive(self, size=1024):
+        with open(self.archive_path, "wb") as f:
+            f.truncate(size)
+
+    def test_it_should_invoke_all_periodic_operations(self):
+        invoked_operations = []
+
+        with _create_collect_logs_handler() as collect_logs_handler:
+            def mock_run(self):
+                invoked_operations.append(self._name)
+
+            with patch.object(PeriodicOperation, "run", side_effect=mock_run, spec=CollectLogsHandler.run):
+                collect_logs_handler.run_and_wait()
+
+                expected_operations = ["collect_and_send_logs"]
+
+                self.assertEqual(invoked_operations.sort(), expected_operations.sort(),
+                                 "The collect logs thread did not invoke the expected operations")
+
+    @patch("azurelinuxagent.ga.collect_logs.CollectLogsHandler.collect_and_send_logs")
+    def test_it_should_only_collect_logs_if_conditions_are_met(self, patch_collect_and_send_logs):
+        # In order to collect logs, three conditions have to be met:
+        # 1) the flag must be set in the conf file
+        # 2) systemd must be managing services
+        # 3) python version 2.7+ which is automatically true for these tests since they are disabled on py2.6
+
+        with _create_collect_logs_handler() as collect_logs_handler:
+            with patch("azurelinuxagent.ga.collect_logs.os.path.exists", return_value=False):
```

Comment:
I'm worried about the scope of this mock; if any other code branches happen to call `os.path.exists` on an unrelated file, could this patch cause issues?

Maybe making a inner function that only special cases the exact file we're looking for, and passing the rest on to the original function?

Reply:
The scope of the mock is just the `collect_logs` module. So, if any other code branches outside of that module hits `os.path.exists`, it would return the actual function, not the mock. However, I like your suggestion of scoping down the mock to the specific file, it will be clearer what the mock is trying to achieve. I will make the change, thanks!